File: python/src/onemin/_base_client.py
"""Base HTTP client for the 1min.ai SDK.

Provides authentication, retry logic with exponential backoff and jitter,
per-request timeout overrides, connection pooling via a single shared httpx.Client
instance, and API key redaction in all public representations.
"""

import os
import random
import time
from typing import Any

import httpx

from onemin._constants import (
    API_KEY_HEADER,
    BASE_URL,
    DEFAULT_BASE_DELAY,
    DEFAULT_MAX_RETRIES,
    DEFAULT_TIMEOUT,
    RETRYABLE_STATUS_CODES,
)
from onemin._exceptions import (
    APIError,
    AuthenticationError,
    BadRequestError,
    ConnectionError,
    InternalServerError,
    NotFoundError,
    RateLimitError,
    TimeoutError,
)


class BaseOneMinClient:
    """Base class providing HTTP transport with auth, retry, timeout, and connection pooling.

    Args:
        api_key: 1min.ai API key. Falls back to ONEMIN_API_KEY environment variable.
        base_url: Override the default API base URL.
        timeout: Default request timeout in seconds (default: 30).
        max_retries: Number of retries for retryable failures (default: 2).
        base_delay: Base delay in seconds for exponential backoff (default: 0.5).

    Raises:
        AuthenticationError: If no API key is provided and ONEMIN_API_KEY is not set.
    """

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        base_delay: float = DEFAULT_BASE_DELAY,
    ) -> None:
        # Auth resolution per D-06: constructor param > env var. No config file.
        resolved = api_key or os.environ.get("ONEMIN_API_KEY")
        if resolved is None:
            raise AuthenticationError(
                "No API key provided. Either pass api_key='...' to OneMinClient() "
                "or set the ONEMIN_API_KEY environment variable.",
                status_code=401,
            )
        self._api_key: str = resolved
        self._base_url: str = (base_url or BASE_URL).rstrip("/")
        self.timeout: float = timeout
        self.max_retries: int = max_retries
        self.base_delay: float = base_delay

        # Single shared httpx.Client instance for connection pooling (INFRA-08)
        self._client = httpx.Client(timeout=httpx.Timeout(timeout))

    def _mask_key(self) -> str:
        """Return a masked version of the API key for display purposes (per INFRA-09).

        Keys longer than 8 chars: show first 4 and last 4 chars with '...' in the middle.
        Shorter keys: return '***'.
        """
        key = self._api_key
        if len(key) > 8:
            return f"{key[:4]}...{key[-4:]}"
        return "***"

    def __repr__(self) -> str:
        """Return masked representation — API key is never exposed in full."""
        return f"OneMinClient(api_key='{self._mask_key()}', base_url='{self._base_url}')"

    def __str__(self) -> str:
        """Return same masked representation as __repr__."""
        return self.__repr__()

    def close(self) -> None:
        """Close the underlying httpx.Client and release connection pool resources."""
        self._client.close()

    def __enter__(self) -> "BaseOneMinClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def _should_retry(self, status_code: int | None, attempt: int) -> bool:
        """Determine whether the request should be retried.

        Args:
            status_code: HTTP status code of the response, or None for network errors.
            attempt: Current attempt number (0-based).

        Returns:
            True if the request should be retried, False otherwise.
        """
        if attempt >= self.max_retries:
            return False
        if status_code is None:
            # Network-level error — retry
            return True
        return status_code in RETRYABLE_STATUS_CODES

    def _get_retry_delay(
        self,
        response: httpx.Response | None,
        attempt: int,
    ) -> float:
        """Calculate the delay before the next retry attempt.

        Honors the Retry-After header for 429 responses (per D-12).
        Otherwise uses exponential backoff with 30-50% jitter.

        Args:
            response: The HTTP response from the last attempt, or None.
            attempt: Current attempt number (0-based), used for backoff exponent.

        Returns:
            Number of seconds to sleep before the next attempt.
        """
        if response is not None and response.status_code == 429:
            retry_after = response.headers.get("Retry-After")
            if retry_after is not None:
                try:
                    return float(retry_after)
                except ValueError:
                    pass

        delay = self.base_delay * (2 ** attempt)
        jitter = delay * random.uniform(0.3, 0.5)
        return delay + jitter

    def _handle_response(self, response: httpx.Response) -> dict[str, Any]:
        """Map HTTP responses to typed exceptions or return parsed JSON.

        API key is never included in error messages (per D-10).

        Args:
            response: The HTTP response to process.

        Returns:
            Parsed JSON response body as a dict.

        Raises:
            AuthenticationError: For 401 responses.
            RateLimitError: For 429 responses.
            NotFoundError: For 404 responses.
            BadRequestError: For 400 responses.
            InternalServerError: For 5xx responses.
            APIError: For other non-success responses (including 403).
        """
        request_id: str | None = response.headers.get("x-request-id")

        if response.is_success:
            try:
                return response.json()
            except Exception:
                raise APIError(
                    f"Non-JSON response: {response.text[:200]}",
                    response.status_code,
                    request_id,
                )

        status = response.status_code

        if status == 401:
            raise AuthenticationError(response.text, 401, request_id)
        elif status == 403:
            # Forbidden — raise APIError, NOT retried per D-11
            raise APIError(response.text, 403, request_id)
        elif status == 429:
            raise RateLimitError(response.text, 429, request_id)
        elif status == 404:
            raise NotFoundError(response.text, 404, request_id)
        elif status == 400:
            raise BadRequestError(response.text, 400, request_id)
        elif status >= 500:
            raise InternalServerError(response.text, status, request_id)
        else:
            raise APIError(response.text, status, request_id)

    def _request(
        self,
        method: str,
        path: str,
        *,
        timeout: float | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Execute an HTTP request with retry logic and error mapping.

        Args:
            method: HTTP method (e.g., 'POST', 'GET').
            path: API path (e.g., '/api/features'). Leading slash is normalized.
            timeout: Per-request timeout override in seconds. If None, uses self.timeout.
                     Domain resources use this to pass their domain-specific timeout
                     (e.g., 90s for image, 300s for video) per INFRA-07.
            **kwargs: Additional arguments passed to httpx.Client.request().

        Returns:
            Parsed JSON response as a dict.

        Raises:
            TimeoutError: If the request times out after all retry attempts.
            ConnectionError: For network-level connection failures.
            APIError (or subclass): For HTTP error responses after all retry attempts.
        """
        url = f"{self._base_url}/{path.lstrip('/')}"
        effective_timeout = timeout if timeout is not None else self.timeout

        # Build headers: merge API key header with any caller-provided headers
        base_headers: dict[str, str] = {
            API_KEY_HEADER: self._api_key,
            "Content-Type": "application/json",
        }
        extra_headers = kwargs.pop("headers", {})
        headers = {**base_headers, **extra_headers}

        last_response: httpx.Response | None = None
        total_attempts = self.max_retries + 1

        for attempt in range(total_attempts):
            if attempt > 0:
                delay = self._get_retry_delay(last_response, attempt - 1)
                time.sleep(delay)

            try:
                response = self._client.request(
                    method,
                    url,
                    headers=headers,
                    timeout=httpx.Timeout(effective_timeout),
                    **kwargs,
                )
            except httpx.TimeoutException as exc:
                if attempt < self.max_retries:
                    last_response = None
                    continue
                raise TimeoutError(str(exc)) from exc
            except httpx.ConnectError as exc:
                # Connection errors are not retried (can't recover from refused connection)
                raise ConnectionError(str(exc)) from exc
            except httpx.RequestError as exc:
                raise ConnectionError(str(exc)) from exc

            if self._should_retry(response.status_code, attempt):
                last_response = response
                continue

            return self._handle_response(response)

        # Exhausted all attempts — handle the last response
        return self._handle_response(last_response)  # type: ignore[arg-type]

File: typescript/src/base-client.ts
import {
  AuthenticationError,
  APIError,
  RateLimitError,
  NotFoundError,
  BadRequestError,
  InternalServerError,
  ConnectionError,
  TimeoutError,
  UnsupportedModelError,
} from './error.js';
import {
  BASE_URL,
  DEFAULT_TIMEOUT,
  DEFAULT_MAX_RETRIES,
  DEFAULT_BASE_DELAY,
  RETRYABLE_STATUS_CODES,
  API_KEY_HEADER,
} from './constants.js';
import { MODEL_CATALOGUE } from './models-data.js';

const UNSUPPORTED_MODEL_NAME_RE = /Model\s+(\S+?)\s+is not supported/i;
const MAX_SUGGESTIONS = 6;

function domainForModel(model: string): string | null {
  for (const [domain, entries] of Object.entries(MODEL_CATALOGUE)) {
    for (const entry of entries) {
      if (entry.id === model || entry.constant === model) return domain;
    }
  }
  return null;
}

function suggestModels(requested: string | null): string[] {
  const domain = (requested ? domainForModel(requested) : null) ?? 'text';
  return (MODEL_CATALOGUE[domain] ?? []).slice(0, MAX_SUGGESTIONS).map((m) => m.id);
}

function maybeUnsupportedModelError(
  body: string,
  status: number,
  requestId: string | undefined,
): UnsupportedModelError | null {
  let payload: unknown;
  try {
    payload = JSON.parse(body);
  } catch {
    return null;
  }
  if (!payload || typeof payload !== 'object') return null;
  const obj = payload as Record<string, unknown>;
  if (obj.errorCode !== 'UNSUPPORTED_MODEL') return null;
  const apiMessage = typeof obj.message === 'string' ? obj.message : '';
  const match = apiMessage.match(UNSUPPORTED_MODEL_NAME_RE);
  const requested = match ? match[1] : null;
  const suggestions = suggestModels(requested);
  const suggestionText = suggestions.length
    ? ` Try one of: ${suggestions.join(', ')}.`
    : '';
  const msg =
    `${apiMessage} Edit data/models.json and run scripts/sync_models.py if you've added a new model.${suggestionText}`;
  return new UnsupportedModelError(msg, status, requestId, requested, suggestions);
}

export interface ClientOptions {
  apiKey?: string;
  baseUrl?: string;
  timeout?: number;     // milliseconds, default 30000
  maxRetries?: number;  // default 2
  baseDelay?: number;   // milliseconds, default 500
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class BaseOneMinClient {
  protected readonly apiKey: string;
  readonly baseUrl: string;
  readonly timeout: number;
  readonly maxRetries: number;
  readonly baseDelay: number;

  constructor(options: ClientOptions = {}) {
    const key = options.apiKey ?? process.env.ONEMIN_API_KEY;
    if (!key) {
      throw new AuthenticationError(
        'No API key provided. Either pass apiKey to OneMinClient() or set the ONEMIN_API_KEY environment variable.',
        401,
      );
    }
    this.apiKey = key;
    this.baseUrl = options.baseUrl ?? BASE_URL;
    this.timeout = options.timeout ?? DEFAULT_TIMEOUT;
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.baseDelay = options.baseDelay ?? DEFAULT_BASE_DELAY;
  }

  toString(): string {
    const key = this.apiKey;
    const masked = key.length > 8
      ? `${key.slice(0, 4)}...${key.slice(-4)}`
      : '***';
    return `OneMinClient(apiKey='${masked}', baseUrl='${this.baseUrl}')`;
  }

  protected isRetryable(statusCode: number): boolean {
    return RETRYABLE_STATUS_CODES.has(statusCode);
  }

  protected getRetryDelay(
    statusCode: number | null,
    headers: Headers | null,
    attempt: number,
  ): number {
    // Honor Retry-After header for 429 responses
    if (statusCode === 429 && headers) {
      const retryAfter = headers.get('Retry-After');
      if (retryAfter) {
        const parsed = parseFloat(retryAfter);
        if (!isNaN(parsed) && parsed > 0) {
          // If Retry-After is in seconds (common), convert to ms
          // Values > 1000 are assumed to already be ms; values <= 1000 are seconds
          return parsed <= 1000 ? parsed * 1000 : parsed;
        }
      }
    }
    const delay = this.baseDelay * Math.pow(2, attempt);
    const jitter = delay * (0.3 + Math.random() * 0.2);
    return delay + jitter;
  }

  protected async handleResponse<T>(response: Response): Promise<T> {
    const requestId = response.headers.get('x-request-id') ?? undefined;
    const status = response.status;

    if (response.ok) {
      return response.json() as Promise<T>;
    }

    let body: string;
    try {
      body = await response.text();
    } catch {
      body = `HTTP ${status} error`;
    }

    switch (status) {
      case 401:
        throw new AuthenticationError(body, 401, requestId);
      case 403:
        throw new APIError(body, 403, requestId);
      case 429:
        throw new RateLimitError(body, 429, requestId);
      case 404:
        throw new NotFoundError(body, 404, requestId);
      case 400: {
        const unsupported = maybeUnsupportedModelError(body, 400, requestId);
        if (unsupported) throw unsupported;
        throw new BadRequestError(body, 400, requestId);
      }
      default:
        if (status >= 500) {
          throw new InternalServerError(body, status, requestId);
        }
        throw new APIError(body, status, requestId);
    }
  }

  protected async request<T>(
    method: string,
    path: string,
    body?: unknown,
    options?: { timeout?: number },
  ): Promise<T> {
    const url = `${this.baseUrl}/${path.replace(/^\//, '')}`;
    const effectiveTimeout = options?.timeout ?? this.timeout;

    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0 && lastError !== undefined) {
        // Determine retry-after delay based on previous response status (if available)
        const statusCode = lastError instanceof APIError ? lastError.statusCode : null;
        const delay = this.getRetryDelay(statusCode, null, attempt - 1);
        if (delay > 0) {
          await sleep(delay);
        }
      }

      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), effectiveTimeout);

      try {
        const fetchOptions: RequestInit = {
          method,
          headers: {
            [API_KEY_HEADER]: this.apiKey,
            'Content-Type': 'application/json',
          },
          signal: ctrl.signal,
        };

        if (body !== undefined) {
          (fetchOptions as RequestInit & { body: string }).body = JSON.stringify(body);
        }

        let response: Response;
        try {
          response = await fetch(url, fetchOptions);
        } catch (err) {
          // Handle abort (timeout) vs network error
          if (err instanceof DOMException && err.name === 'AbortError') {
            throw new TimeoutError('Request timed out');
          }
          if (err instanceof TypeError) {
            // Network error — retry if possible
            if (attempt < this.maxRetries) {
              lastError = new ConnectionError((err as TypeError).message);
              continue;
            }
            throw new ConnectionError((err as TypeError).message);
          }
          throw err;
        }

        if (!response.ok && this.isRetryable(response.status) && attempt < this.maxRetries) {
          // Clone response to read it later
          lastError = new APIError(
            `Retryable error: ${response.status}`,
            response.status,
          );
          continue;
        }

        return await this.handleResponse<T>(response);
      } finally {
        clearTimeout(timer);
      }
    }

    // Should never reach here but TypeScript needs this
    throw lastError ?? new ConnectionError('Request failed after retries');
  }

  public _request<T>(
    method: string,
    path: string,
    body?: unknown,
    options?: { timeout?: number },
  ): Promise<T> {
    return this.request<T>(method, path, body, options);
  }
}
